### PR TITLE
Issue 4579 - Unconsumed attributes warning&error behavior

### PR DESCRIPTION
**Pull request by arno-lunarg**
Fix for https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/4579

Completely unconsumed attributes are allowed, and should result in a performance warning if the best practices layer is enabled.
Partially consumed attributes are allowed if `maintenance4` is enabled, but should still result in a performance warning. They are illegal when `maintenance4` is disabled, and result in an error.

### CODE REVIEW

**Review by juan-lunarg — CHANGES_REQUESTED** [review]

File: layers/best_practices_utils.cpp [hunk 1]
```diff
@@ -1095,6 +1104,97 @@ void BestPractices::ManualPostCallRecordCreateComputePipelines(VkDevice device,
...
+    while (output_it != shader_outputs.end() || input_it != shader_inputs.end()) {
+        const bool output_at_end = output_it == shader_outputs.end();
+        const bool input_at_end = input_it == shader_inputs.end();
+
+        ShaderAttribute out_attrib;
+        out_attrib.is_valid = !output_at_end;
+        if (out_attrib.is_valid) {
+            out_attrib.location = output_it->first.first;
+            out_attrib.written_components_count = producer.GetNumComponentsInBaseType(producer.get_def(output_it->second.type_id));
+        }
```
Comment (ncesario-lunarg) [inline layers/best_practices_utils.cpp:1145]: Looks like `out_attrib` is read-only in this function? Might make sense to put this in a constructor and make `out_attrib` `const`, but definitely not necessary. Same comment for `in_attrib` below (might be able to consolidate that initialization code a bit as well then).

File: layers/best_practices_utils.cpp [hunk 2]
```diff
@@ -1113,6 +1213,33 @@ bool BestPractices::PreCallValidateCreateGraphicsPipelines(VkDevice device, VkPi
...
+                if (consumer.module_state->has_valid_spirv && producer.module_state->has_valid_spirv) {
+                    auto producer_id = GetShaderStageId(producer.stage_flag);
+                    auto consumer_id = GetShaderStageId(consumer.stage_flag);
+                    skip |= ValidateInterfaceBetweenStages(*producer.module_state.get(), producer.entrypoint,
+                                                           &shader_stage_attribs[producer_id], *consumer.module_state.get(),
+                                                           consumer.entrypoint, &shader_stage_attribs[consumer_id]);
+                }
+            }
+        }
+    }
```
Comment (ncesario-lunarg) [inline layers/best_practices_utils.cpp:1241]: Not a ton of code here, but another duplication of CoreChecks code. Again, might be more effort than it's worth to consolidate, but if we can, that would be good.

**Review by ncesario-lunarg — CHANGES_REQUESTED** [review]
Overall this change LGTM, and I appreciate the comments :)

Just a lot of duplication that would be nice to consolidate. Some of it I think we can consolidate pretty easily, but other parts may be more effort than it's worth.

**Comment by arno-lunarg** [comment]
I am going to improve on duplication, thank you for the comments @juan-lunarg and @ncesario-lunarg!

File: layers/best_practices_utils.cpp [hunk 3]
```diff
@@ -1095,6 +1104,97 @@ void BestPractices::ManualPostCallRecordCreateComputePipelines(VkDevice device,
     pipeline_cache_ = pipelineCache;
 }
 
+static uint32_t GetShaderStageId(VkShaderStageFlagBits stage) {
+    uint32_t bit_pos = uint32_t(u_ffs(stage));
+    return bit_pos - 1;
+}
```
[the remark this answers is not in the document]
Comment (jeremyg-lunarg) [reply layers/best_practices_utils.cpp:1110]: This is already defined in shader_validation.cpp, it should be made available outside that file somehow.

**Comment by arno-lunarg** [comment]
Ok so I refactored `CoreChecks::ValidateInterfaceBetweenStages` and `BestPractices::ValidateInterfaceBetweenStages`, As well as the for loop you mentioned [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/4630#discussion_r989556763) @ncesario-lunarg.
It took way more time than anticipated, because of how I wanted to refactor the ValidateInterface functions. I wanted to separate the iterators increments part from the checks part, so that those two functions only need to worry about checking a (out, in) attribute pair. Turns out redoing iterators increments was kind of hard. Also, the check parts now heavily rely on the `ShaderAttribute` struct, it makes the code simpler and more explicit.

Some bugs appeared along the way, and some still are very much alive, because I did not know about the `component` attribute layout qualifier. This keyword is a nightmare, especially when combined with the special rules for types like `dvec3` and `dvec4`. ATM, I still do not handle correctly those types, since I do not know how to see if an attribute has such a type?

**Comment by arno-lunarg** [comment]
Failing test is `VkLayerTest.PipelineInUseDestroyedSignaled` on tcwinnvi2 debug 64, seems unrelated to those changes

File: layers/shader_validation.cpp [hunk 4]
```diff
@@ -3146,108 +3140,55 @@ bool CoreChecks::ValidateInterfaceBetweenStages(const SHADER_MODULE_STATE &produ
...
-            if ((input_first.first > output_first.first) || input_at_end || (output_component + 1 == output_length)) {
-                output_it++;
-                output_component = 0;
-            } else {
-                output_component++;
+                std::string msg = std::string{producer_stage->name} + " writes to output location " + std::to_string(out.location) +
+                                  '.' + std::to_string(out.component) + " which is not consumed by " + consumer_stage->name +
+                                  ". Enable VK_KHR_maintenance4 device extension to allow relaxed interface matching between "
+                                  "input and output vectors. ";
+                skip |= LogError(producer.vk_shader_module(), kVUID_Core_Shader_OutputNotConsumed, "%s", msg.c_str());
```
Comment (spencer-lunarg) [inline layers/shader_validation.cpp:3154]: I think this is `VUID-RuntimeSpirv-OpTypeVector-06816`/`VUID-RuntimeSpirv-OpTypeVector-06817`

>    Any OpTypeVector output interface variables must: not have a higher
    Component Count than a matching OpTypeVector input
    interface variable
    
so would make this

```cpp
const char* vuid = IsExtEnabled(device_extensions.vk_khr_maintenance4) ? "VUID-RuntimeSpirv-OpTypeVector-06817" : "VUID-RuntimeSpirv-OpTypeVector-06816";
```

(and then apply to the test case below to check for this)

**Comment by arno-lunarg** [comment]
Closed for now, spotted few more problems related to 64 bits types, I have to make some changes.

**Pull request closed without merging**